src/proto/batch/mod.rs

```diff
@@ -0,0 +1,338 @@
+/// ```no_run
+/// # use faktory::Error;
+/// # use faktory::{Producer, Job, Batch, Tracker};
+/// let mut prod = Producer::connect(None)?;
+/// let job = Job::builder("image").build();
+/// let cb_job = Job::builder("clean_up").build();
+/// let b = Batch::builder("Description...".to_string()).with_complete_callback(cb_job);
+///
+/// let mut b = prod.start_batch(b)?;
+/// let bid = b.id().to_string();
+/// b.add(job)?;
+/// b.commit()?;
+///
+/// let mut t = Tracker::connect(None).unwrap();
+/// let s = t.get_batch_status(bid).unwrap().unwrap();
+/// assert_eq!(s.total, 1);
+/// assert_eq!(s.pending, 1);
+/// assert_eq!(s.description, Some("Description...".into()));
+/// assert_eq!(s.complete_callback_state, ""); // has not been queued;
+/// # Ok::<(), Error>(())
+/// ```
+#[derive(Serialize, Debug, Builder)]
+#[builder(
+    custom_constructor,
+    setter(into),
+    build_fn(name = "try_build", private)
+)]
+pub struct Batch {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(skip))]
+    parent_bid: Option<String>,
+
+    /// Batch description for Faktory WEB UI.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// On success callback.
+    ///
+    /// This job will be queued by the Faktory server provided
+    /// all the jobs belonging to this batch have been executed successfully.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) success: Option<Job>,
+
+    /// On complete callback.
+    ///
+    /// This job will be queue by the Faktory server after all the jobs
+    /// belonging to this batch have been executed, even if one/some/all
+    /// of the workers have failed.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    #[builder(setter(custom))]
+    pub(crate) complete: Option<Job>,
+}
+
+impl Batch {
+    /// Create a new `BatchBuilder`.
+    pub fn builder(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder::new(description)
+    }
+}
+
+impl BatchBuilder {
+    fn build(&mut self) -> Batch {
+        self.try_build()
+            .expect("All required fields have been set.")
+    }
+
+    /// Create a new `BatchBuilder` with optional description of the batch.
+    pub fn new(description: impl Into<Option<String>>) -> BatchBuilder {
+        BatchBuilder {
+            description: Some(description.into()),
+            ..Self::create_empty()
+        }
+    }
+
+    fn success(&mut self, success_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.success = Some(success_cb.into());
+        self
+    }
+
+    fn complete(&mut self, complete_cb: impl Into<Option<Job>>) -> &mut Self {
+        self.complete = Some(complete_cb.into());
+        self
+    }
+
+    /// Create a `Batch` with only `success` callback specified.
+    pub fn with_success_callback(&mut self, success_cb: Job) -> Batch {
```

Comment:
I'd probably make this and the other methods take `self` so that they don't have to clone all the fields. Callers can always choose to `.clone()` before calling the method if they want to reuse the same builder values.

Reply:
There is a [config option](https://docs.rs/derive_builder/latest/derive_builder/#owned-aka-consuming) in `batch_builder` precisely for that.

Reply:
```rust
let b = BatchBuilder::new().description("Batch description");
let _batch_with_complete_cb = b.clone().with_complete_callback(Job::builder("jt").build());
let _batch_with_success_cb = b.with_success_callback(Job::builder("jt").build());
```